### PR TITLE
boards/opencm9-04 : initial support

### PR DESCRIPTION
**Pull request by astralien3000**
Hello !

This is just a work in progress, because I am currently only able to make the custom FLASHER work.
I am not even able to make a led blink...

Here are the best information I can get on this board : 
http://support.robotis.com/en/product/controller/opencm9.04.htm

### CODE REVIEW

**Comment by aabadie** [comment]
Thanks, this board looks great.

Your port is still missing a `board.c` and an include directory with a `board.h` and `periph_conf.h`. Take example on `nucleo-f103` board  as they share a similar CPU.

I guess you can keep the same CLOCK definition macros and, from the schematic, configure 1 or 2 UARTs (PA2/PA3 and PA9/PA10). For the user LED, I don't see what's the pin mapping.

**Comment by astralien3000** [comment]
According to the [schematic](http://support.robotis.com/en/baggage_files/opencm/opencm904_rev_10_final_schematic.pdf), the STATUS_LED(GREEN) should be linked with D14, or PB9.
If I link GND and D14 (with an external wire), the led is ON, and OFF otherwise.
I think I configured correctly LED0 in board.h, but if I try to call `LED0_ON;` or `LED0_OFF;`, nothing changes.

**Comment by astralien3000** [comment]
The LED problem seemed to be a linker script broblem. Indeed, the bootloader need the program to begin at an other address thant the default one.
So, I am able to turn ON and OFF the led in the board_init function, but not in main.
Also, as I understand, the "USB serial" is not an UART for the microcontroller, so that, It needs the full USB stack to use this serial.

**Comment by astralien3000** [comment]
Here is my problem : I can't run the main function because the program has a problem before getting to the main.
I know that I can run the functions : 
 - [board_init](https://github.com/astralien3000/RIOT/blob/opencm9-04/boards/opencm9-04/board.c#L24)
 - [kernel_init](https://github.com/RIOT-OS/RIOT/blob/master/core/kernel_init.c#L81)

And I reach the function [thread_arch_start_threading](https://github.com/RIOT-OS/RIOT/blob/master/cpu/cortexm_common/thread_arch.c#L281) which triggers an "svc" interrupt.

And then I don't know where go the execution thread.
As far as I understand, the [isr_svc](https://github.com/RIOT-OS/RIOT/blob/master/cpu/cortexm_common/thread_arch.c#L334) function should be executed, but it is not.

I think the [interrupt vector](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f1/vectors.c#L95) is not found by the microcontroller. Indeed, the board has a bootloader by default, and the user programe don't stard at 0x08000000 but at 0x08003000 (I use a custom [ldscript](https://github.com/astralien3000/RIOT/blob/opencm9-04/cpu/stm32f1/ldscripts/stm32f103cb_opencm9-04.ld). So that the vector is not where it should be...

Does someone has an idea on how I can fix this ?

**Comment by aabadie** [comment]
Thanks @astralien3000 for the in-depth diagnosis. Good job !

> Does someone has an idea on how I can fix this ?

This is out of my knowledge but maybe @kaspar030 or @haukepetersen can help ?

**Comment by astralien3000** [comment]
Some good update !
It seems like the error I had was because of [this line](https://github.com/RIOT-OS/RIOT/blob/master/cpu/cortexm_common/cortexm_init.c#L40) : 
```c
SCB->VTOR = CPU_FLASH_BASE;
```

CPU_FLASH_BASE is equal to 0x08000000 not depending on what is in the linker script.

If I replace this line by : 
```c
SCB->VTOR = 0x08003000;
```

I can go to the main !
And UART is working !

But it is dirty, and I don't know what would be the clean way to do the same.
Indeed, `CPU_FLASH_BASE` is a preprocessor defined value, but it should be something like a linker_script defined value. Also, I don't like the name ^^, since we are talking about the vector's address, not the FLASH's begining address.

**Comment by astralien3000** [comment]
Hmm : 
```
2017-01-19 12:20:30,691 - INFO # �main(): This is RIOT! (Version: 2017.01-devel-512-g3895a-223-55.wifi-inria-saclay.saclay.inria.fr-opencm9-04)
2017-01-19 12:20:33,646 - INFO # *** RIOT kernel panic:
2017-01-19 12:20:33,646 - INFO # DUMMY HANDLER
2017-01-19 12:20:33,646 - INFO # 
2017-01-19 12:20:33,647 - INFO # *** halted.
```

Why is the dummy handler triggered since there is no dummy handler in the [stm32f1's vector implementation](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f1/vectors.c) ?

**Comment by aabadie** [comment]
> CPU_FLASH_BASE is equal to 0x08000000 not depending on what is in the linker script.

Better use [this line](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f1/include/stm32f103xb.h#L612) in the cmsis file. Then you need to play with the defines to include the right one.

**Comment by astralien3000** [comment]
I have done like the `spark-core` board (modified the VTOR field in board_init).
The interruption triggered is an USB interruption initialized by the bootloader : 
```
2017-01-19 15:04:08,501 - INFO # �main(): This is RIOT! (Version: 2017.01-devel-513-g34e8f-223-55.wifi-inria-saclay.saclay.inria.fr-opencm9-04)
2017-01-19 15:04:08,502 - INFO # Hello World!
2017-01-19 15:04:08,507 - INFO # You are running RIOT on a(n) opencm9-04 board.
2017-01-19 15:04:08,521 - INFO # This board features a(n) stm32f1 MCU.
2017-01-19 15:04:12,037 - INFO # *** RIOT kernel panic:
2017-01-19 15:04:12,038 - INFO # isr_usb_lp_can1_rx0
2017-01-19 15:04:12,038 - INFO # 
2017-01-19 15:04:12,039 - INFO # *** halted.
```

**Comment by kYc0o** [comment]
Is the bootloader really needed? cannot you overwrite it to start at address 0x08000000 ? What are you using to flash the device?

**Comment by astralien3000** [comment]
I use the bootloader ^^
There is a SWD/JTAG port, but you need an external JTAG to program/debug the board.

**Comment by kYc0o** [comment]
And which kind of bootloader it has? it's not FTDI or CMSIS-DEP based? (like iotlab-m3 or samr21-xpro)

**Comment by astralien3000** [comment]
I think it is a custom one, made by ROBOTIS (the board's vendor).
The USB is not a FTDI, but a pure USB that is manager by the controller.
The bootloader seems to have the USB stack implemented, and use it to flash the board.

I use a custom python program (also made by ROBOTIS) to load the code to the board.

**Comment by astralien3000** [comment]
Good news : I disabled the USB in board_init, and everything seem to work fine.

**Comment by astralien3000** [comment]
> Is the bootloader really needed?

@kYc0o the bootloader is never needed when you have a JTAG, but I had none for the past few days, and it also would oblige people to buy one...

Also, It is possible to handle both configuration : one using the bootloader which need some hacks to work well, and one using the JTAG (2 separate "boards" ?)

**Comment by kYc0o** [comment]
OK I see, then yes you need to change the `CPU_FLASH_BASE ` as you did, be aware of #5636, maybe it can solve this issue. The approach you used is OK but we indeed need a way to abstract this hardcoded address and provide it from somewhere else, for this case and others.

**Comment by astralien3000** [comment]
> be aware of #5636

👍 That would be the way I would implement it ! ^^

Also, I put the linker script in the `cpu/stm32f1` directory, which I think is dirty also (the spark-core board do the same). Is there a better way ?

**Comment by astralien3000** [comment]
Concerning the problem of robotis-loader.py, which has no license header, I added the README from the [author's repository](https://github.com/Gregwar/robotis-loader/), which says that is under MIT.
Is that OK ?

File: boards/opencm9-04/include/board.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * @name xtimer configuration
+ * @{
+ */
+#define XTIMER_WIDTH        (16)
+#define XTIMER_BACKOFF      5
+/** @} */
+
+/**
+ * @name Macros for controlling the on-board LEDs.
```
Comment (aabadie) [inline boards/opencm9-04/include/board.h]: LED: there's only one

**Comment by aabadie** [comment]
Except from a minor cosmetic comment, changes looks good: untested ACK

**Comment by aabadie** [comment]
I forgot: please squash ;)

**Comment by aabadie** [comment]
About the MIT license. I'm not a lawyer, so I might not the right person to answer. Maybe @OlegHahm  or @emmanuelsearch  have an answer ?
Otherwise you can git grep MIT in the repo, perhaps there's already this kind of code inclusion.

**Comment by emmanuelsearch** [comment]
I think the standard way is to insert a MIT license header in the robotis-loader.py file, as well as the name of the author.

File: boards/opencm9-04/Makefile.include [hunk 2]
```diff
@@ -0,0 +1,23 @@
+## the cpu to build for
+export CPU = stm32f1
+export CPU_MODEL = stm32f103cb
+
+# custom linkerscript
+export LINKER_SCRIPT = stm32f103cb_opencm9-04.ld
+
+# custom flasher to ust with the bootloader
```
Comment (aabadie) [inline boards/opencm9-04/Makefile.include]: ust ?

File: boards/opencm9-04/Makefile.include [hunk 3]
```diff
@@ -0,0 +1,23 @@
+## the cpu to build for
```
Comment (aabadie) [inline boards/opencm9-04/Makefile.include]: 2 `##`

File: boards/opencm9-04/board.c [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+#include "board.h"
+#include "periph/gpio.h"
+#include "log.h"
+
+void board_init(void)
+{
+    /* initialize the CPU */
+    cpu_init();
+
+    /* initialize the boards LEDs */
```
Comment (aabadie) [inline boards/opencm9-04/board.c]: I thought you changed LED**s**

File: boards/opencm9-04/dist/robotis-loader.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+# Reading command line
+#if len(sys.argv) != 3:
+#    exit('! Usage: robotis-loader.py <serial-port> <binary>')
+#pgm, port, binary = sys.argv
+
+pgm = sys.argv[0]
+port = os.environ["PORT"]
+binary = os.environ["HEXFILE"]
+
+# Helper to prints a progress bar
```
Comment (aabadie) [inline boards/opencm9-04/dist/robotis-loader.py]: print**s**

Comment (aabadie) [inline boards/opencm9-04/dist/robotis-loader.py]: `that` instead of `to` is better and keep `prints`

File: boards/opencm9-04/include/board.h [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ * @brief       Board specific files for the OpenCM9.04 board
+ * @{
+ *
+ * @file
+ * @brief       Board specific definitions for the OpenCM9.04 board
+ *
+ * @author      Loïc Dauphin <loic.dauphin@inria.fr>
+ */
+
+#ifndef BOARD_H_
```
Comment (aabadie) [inline boards/opencm9-04/include/board.h]: Trailing uinderscores should be removed see #6407

File: boards/opencm9-04/include/periph_conf.h [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+ * @ingroup     boards_opencm9-04
+ * @{
+ *
+ * @file
+ * @brief       Peripheral MCU configuration for the opencm9-04 board
+ *
+ * @author      Loïc Dauphin <loic.dauphin@inria.fr>
+ */
+
+#ifndef PERIPH_CONF_H_
```
Comment (aabadie) [inline boards/opencm9-04/include/periph_conf.h]: trailing underscore

File: boards/opencm9-04/include/periph_conf.h [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+#define I2C_1_EVT_IRQ       I2C2_EV_IRQn
+#define I2C_1_EVT_ISR       isr_i2c2_ev
+#define I2C_1_ERR_IRQ       I2C2_ER_IRQn
+#define I2C_1_ERR_ISR       isr_i2c2_er
+/* I2C 1 pin configuration */
+#define I2C_1_SCL_PIN       GPIO_PIN(PORT_B, 10) /* D1 */
+#define I2C_1_SDA_PIN       GPIO_PIN(PORT_B, 11) /* D0 */
+/** @} */
+
+/**
```
Comment (aabadie) [inline boards/opencm9-04/include/periph_conf.h]: You don't want to use SPI and there's a big PR about that #4780. I propose to remove the SPI definitions from this file and the dependency from the Makefile.deps.

File: boards/opencm9-04/board.c (hunk 4 above)
Comment (astralien3000) [inline boards/opencm9-04/board.c]: I thought I did not miss one ;)

File: boards/opencm9-04/include/board.h [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+#define LED0_TOGGLE         (LED_PORT->ODR ^= LED0_MASK)
+/** @} */
+
+/**
+ * @brief   User button
+ */
+#define BTN_B1_PIN          GPIO_PIN(PORT_C, 15)
+
+/**
+ * @brief Use the USART1 for STDIO on this board
```
Comment (aabadie) [inline boards/opencm9-04/include/board.h]: If you set UART_DEV(0) then, according to the board config, it should be USART2 (not USART1)

Comment (astralien3000) [inline boards/opencm9-04/include/board.h]: Yes, this is what I mean ^^
USART2 is connected to A4 and A5, and to 4-pin communication port (on the robot : the bluetooth module)  

USART1 is connected to the Dynamixel TTL bus.

Comment (astralien3000) [inline boards/opencm9-04/include/board.h]: oups sorry, I did not understood

**Comment by aabadie** [comment]
I have problems flashing the board : the flasher code is a bit a mess and not python 3 compatible...

**Comment by aabadie** [comment]
Ok, I think I fixed it

**Comment by astralien3000** [comment]
Should I modify the script to make it easier to use ? Or is it OK ?

**Comment by aabadie** [comment]
The usage is ok. It's just that it's absolutely not python 3 compatible

**Comment by aabadie** [comment]
@astralien3000 Can you also add some notes about how to flash the board in a README ?

**Comment by astralien3000** [comment]
Is it better "in the code" or the wiki ?

**Comment by aabadie** [comment]
I think both are ok.

I could test the board with success using `examples/default` application : I had the shell. Good job !

**Comment by astralien3000** [comment]
I have added a [HOWTO text in the loader](https://github.com/astralien3000/RIOT/blob/1fb47b5ee78abf495b7d2519672e9f66fe627119/boards/opencm9-04/dist/robotis-loader.py#L45), is it ok ?
Well, after that I think I have squashed correctly, don't I ?

**Pull request merged**